Prototype for export_for_training

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129547
* #129249
* __->__ #129092
* #128077


This PR implements export_for_training where the IR is not-functional, pre-dispatch aten IR. The general strategy:
1. Call dynamo to get torch IR
2. Lift param/buffer 
3. call make_fx 

TODO:
1. run_decomp doesn't work 
2. not-strict is not supported

Differential Revision: [D59069087](https://our.internmc.facebook.com/intern/diff/D59069087)

tugsbayasgalan committed:
Prototype for export_for_training

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Prototype for export_for_training"




[ghstack-poisoned]

SherlockNoMad commented:
Is there a way to get Trainining IR with non-strict mode? i.e. no dynamo.

tugsbayasgalan committed:
Update on "Prototype for export_for_training"



This PR implements export_for_training where the IR is not-functional, pre-dispatch aten IR. The general strategy:
1. Call dynamo to get torch IR
2. Lift param/buffer 
3. call make_fx 

TODO:
1. run_decomp doesn't work 
2. not-strict is not supported


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Prototype for export_for_training"



This PR implements export_for_training where the IR is not-functional, pre-dispatch aten IR. The general strategy:
1. Call dynamo to get torch IR
2. Lift param/buffer 
3. call make_fx 

TODO:
1. run_decomp doesn't work 
2. not-strict is not supported


[ghstack-poisoned]

tugsbayasgalan commented:
> Is there a way to get Trainining IR with non-strict mode? i.e. no dynamo.

There is. But it will be non-trivial work because there are some aot-export-module specific logic that handles buffer re-assignments that @avikchaudhuri  worked on. We will need to port them somehow. I think executorch doesn't do non-strict today, so we might be ok with just not imnplementing it for now

tugsbayasgalan committed:
Update on "Prototype for export_for_training"



This PR implements export_for_training where the IR is not-functional, pre-dispatch aten IR. The general strategy:
1. Call dynamo to get torch IR
2. Lift param/buffer 
3. call make_fx 

TODO:
1. run_decomp doesn't work 
2. not-strict is not supported


[ghstack-poisoned]

SherlockNoMad commented on torch/export/_trace.py:
```diff
@@ -1486,6 +1683,281 @@ def forward(self, *args, **kwargs):
 
+@_log_export_wrapper
+@_disable_prexisiting_fake_mode
+def _export_for_training(
```
This implementation has a lot of duplications with `_strict_export`. 
So I wonder what's the delta? can we reuse code?

SherlockNoMad commented on torch/export/_trace.py:
```diff
@@ -1348,6 +1355,196 @@ def _strict_export(
     )
 
 
+def _export_to_aten_ir_make_fx(
```
I am trying to make sense of the different of this function and '_export_to_aten_ir` 

There are quite some duplication, and I am not sure about the delta.

tugsbayasgalan committed:
Update on "Prototype for export_for_training"



This PR implements export_for_training where the IR is not-functional, pre-dispatch aten IR. The general strategy:
1. Call dynamo to get torch IR
2. Lift param/buffer 
3. call make_fx 

TODO:
1. run_decomp doesn't work 
2. not-strict is not supported


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Prototype for export_for_training"



This PR implements export_for_training where the IR is not-functional, pre-dispatch aten IR. The general strategy:
1. Call dynamo to get torch IR
2. Lift param/buffer 
3. call make_fx 

TODO:
1. run_decomp doesn't work 
2. not-strict is not supported


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Prototype for export_for_training"



This PR implements export_for_training where the IR is not-functional, pre-dispatch aten IR. The general strategy:
1. Call dynamo to get torch IR
2. Lift param/buffer 
3. call make_fx 

TODO:
1. run_decomp doesn't work 
2. not-strict is not supported


[ghstack-poisoned]

zhxchen17 commented on torch/export/_trace.py:
```diff
@@ -1366,6 +1563,281 @@ def forward(self, *args, **kwargs):
+    args: Tuple[Any, ...],
+    kwargs: Optional[Dict[str, Any]] = None,
+    dynamic_shapes: Optional[Union[Dict[str, Any], Tuple[Any], List[Any]]] = None,
+    *,
```
@tugsbayasgalan could you leave a TODO for tensor conversion .to() might not be preserved during export.

tugsbayasgalan committed:
Update on "Prototype for export_for_training"



This PR implements export_for_training where the IR is not-functional, pre-dispatch aten IR. The general strategy:
1. Call dynamo to get torch IR
2. Lift param/buffer 
3. call make_fx 

TODO:
1. run_decomp doesn't work 
2. not-strict is not supported


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Prototype for export_for_training"



This PR implements export_for_training where the IR is not-functional, pre-dispatch aten IR. The general strategy:
1. Call dynamo to get torch IR
2. Lift param/buffer 
3. call make_fx 

TODO:
1. run_decomp doesn't work 
2. not-strict is not supported


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Prototype for export_for_training"



This PR implements export_for_training where the IR is not-functional, pre-dispatch aten IR. The general strategy:
1. Call dynamo to get torch IR
2. Lift param/buffer 
3. call make_fx 

TODO:
1. run_decomp doesn't work 
2. not-strict is not supported


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Prototype for export_for_training"



This PR implements export_for_training where the IR is not-functional, pre-dispatch aten IR. The general strategy:
1. Call dynamo to get torch IR
2. Lift param/buffer 
3. call make_fx 

TODO:
1. run_decomp doesn't work 
2. not-strict is not supported


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Prototype for export_for_training"



This PR implements export_for_training where the IR is not-functional, pre-dispatch aten IR. The general strategy:
1. Call dynamo to get torch IR
2. Lift param/buffer 
3. call make_fx 

TODO:
1. run_decomp doesn't work 
2. not-strict is not supported


[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Prototype for export_for_training"



This PR implements export_for_training where the IR is not-functional, pre-dispatch aten IR. The general strategy:
1. Call dynamo to get torch IR
2. Lift param/buffer 
3. call make_fx 

TODO:
1. run_decomp doesn't work 
2. not-strict is not supported


[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59069087).

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59069087).

zhxchen17 commented on torch/export/_trace.py:
```diff
@@ -1104,6 +1111,37 @@ def _strict_export(
+    )
+
+
+def _generic_export_lower_to_aten_callback(
```
```suggestion
def _strict_export_lower_to_aten_ir(
```

tugsbayasgalan committed:
Update on "Prototype for export_for_training"



This PR implements export_for_training where the IR is not-functional, pre-dispatch aten IR. The general strategy:
1. Call dynamo to get torch IR
2. Lift param/buffer 
3. call make_fx 

TODO:
1. run_decomp doesn't work 
2. not-strict is not supported

Differential Revision: [D59069087](https://our.internmc.facebook.com/intern/diff/D59069087)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59069087).
